Add basic documents and charm terraform module (#14)
canonical/opencti-operator: docs/explanation/charm-architecture.md

entrypoint.

Pebble is a lightweight, API-driven process supervisor that is responsible for
configuring processes to run in a container and controlling those
processes throughout the workload lifecycle.

Pebble `services` are configured through [layers](https://github.com/canonical/pebble#layer-specification), 
and the following containers represent each one a layer forming the effective 
Pebble configuration, or `plan`:

1. A `charm` container, which contains the charm code that interact with the juju system.
2. The [`opencti`](https://filigran.io/solutions/open-cti/) container itself, which runs the OpenCTI platform and OpenCTI workers.

As a result, if you run a `kubectl get pods` on a namespace named for the Juju 
model you've deployed the OpenCTI charm into, you'll see something like the 
following:

```bash
NAME                                   READY   STATUS    RESTARTS   AGE
opencti-0                              2/2     Running   4 (11m ago)   3h20m
```

And if you run `kubectl describe pod opencti-0`, all the containers will have 
as Command ```/charm/bin/pebble```. That's because Pebble is responsible for
the processes startup as explained above.

> ## Charm Architecture Diagrams

Reviewer: ```suggestion
## Charm architecture diagram
```

```mermaid
C4Container
    System_Ext(k8s-ingress, "Kubernetes Ingress")
    System_Ext(s3, "S3 Compatible Storage")
    Container_Boundary(k8s_model, "K8s Model") {
        Container(ingress, "Nginx Ingress Integrator")
        Container(s3-integrator, "S3 Integrator")
        Container(redis, "Redis")
        Container(opencti, "OpenCTI")
        Container(opencti-connector-1, "OpenCTI Connectors")
        Container(opencti-connector-2, "OpenCTI Connectors")
        Rel(redis, opencti, "redis")
        Rel(ingress, opencti, "ingress")
        Rel(s3-integrator, opencti , "s3")
        BiRel(opencti, opencti-connector-1, "opencti-connector")
        BiRel(opencti, opencti-connector-2, "opencti-connector")
    }
    Container_Boundary(vm_model, "VM Model") {
        Container(rabbitmq, "RabbitMQ")
        Container(opensearch, "OpenSearch")
    }
    Rel(opensearch, opencti, "opensearch")
    Rel(rabbitmq, opencti, "amqp")
    Rel(ingress, k8s-ingress, "control")
    Rel(s3, s3-integrator, "obtain information")

> ```

Reviewer: This diagram is great! It gives a high-level overview of the complex OpenCTI ecosystem. I think it will add a lot of value to the docs! 

I think that this diagram should be under a heading like `## High-level overview of OpenCTI` and should also include a small paragraph describing the different components -- the VM model, the K8s model, why there are only two connectors included in the diagram, etc.

Additionally, there needs to be a diagram under the heading `## Charm architecture diagram` that's specific to the architecture of the OpenCTI charm itself. What's going on internally with the OpenCTI charm? How is the charm containerized?

## OCI images

> We use [Rockcraft](https://canonical-rockcraft.readthedocs-hosted.com/en/latest/) to build OCI Images for <charm-name>. 

Reviewer: ```suggestion
We use [Rockcraft](https://canonical-rockcraft.readthedocs-hosted.com/en/latest/) to build OCI Images for the OpenCTI charm. 
```

> The images are defined in [opencti rock](https://github.com/canonical/opencti-operator/blob/main/opencti_rock/rockcraft.yaml).

Reviewer: ```suggestion
The images are defined in the [opencti rock](https://github.com/canonical/opencti-operator/blob/main/opencti_rock/rockcraft.yaml).
```

### `opencti` Container

The `opencti` container runs the OpenCTI platform and OpenCTI workers.
In every container, one instance of OpenCTI platform and three instances of 
OpenCTI workers are running. The ratio is recommended by the [OpenCTI deployment guide](https://docs.opencti.io/latest/deployment/clustering/).

The workload that this container is running is defined in the [opencti rock](https://github.com/canonical/opencti-operator/blob/main/opencti_rock/rockcraft.yaml).

## Metrics

### OpenCTI metrics

OpenCTI platform in the OpenCTI charm is [configured to expose prometheus on port `14269`](https://docs.opencti.io/latest/deployment/configuration/#telemetry). 
 
## Juju events

For this charm, the following Juju events are observed:

1. `config-changed`
2. `upgrade-charm`
3. `update-status`
4. `secret-changed`
5. `opencti-pebble-ready`
6. `opencti-peer-relation-created`
7. `opencti-peer-relation-changed`
8. `opencti-peer-relation-departed`
9. `opencti-peer-relation-broken`
10. `opencti-pebble-custom-notice`
11. `opencti-connector-relation-joined`
12. `opencti-connector-relation-changed`

> 13. `stop`

Reviewer: Could you offer a brief explanation about what each event is? Or could you link to the Juju docs where these events are discussed?

> And many events observed by charm libraries.

Reviewer: This sentence is a bit ambiguous to me. Are the events listed above observed by charm libraries?

Author: Gotcha, updated, thanks!

> > See more in the Juju docs: [Event](https://juju.is/docs/sdk/event)

Reviewer: ```suggestion
> See more in the Juju docs: [Hook](https://juju.is/docs/sdk/event)
```

## Charm code overview

The `src/charm.py` is the default entry point for a charm and has the 
`OpenCTICharm` Python class which inherits from CharmBase. CharmBase is the 
base class from which all Charms are formed, defined by [Ops](https://juju.is/docs/sdk/ops) 
(Python framework for developing charms).

> > See more in the Juju docs: [Charm](https://juju.is/docs/sdk/constructs#heading--charm)

Reviewer: ```suggestion
> See more in the Juju docs: [Charm](https://canonical-juju.readthedocs-hosted.com/en/latest/user/reference/charm/)
```

configuring processes to run in a container and controlling those
processes throughout the workload lifecycle.

Pebble `services` are configured through [layers](https://github.com/canonical/pebble#layer-specification), 
and the following containers represent each one a layer forming the effective 
Pebble configuration, or `plan`:

1. A `charm` container, which contains the charm code that interact with the juju system.
2. The [`opencti`](https://filigran.io/solutions/open-cti/) container itself, which runs the OpenCTI platform and OpenCTI workers.

As a result, if you run a `kubectl get pods` on a namespace named for the Juju 
model you've deployed the OpenCTI charm into, you'll see something like the 
following:

```bash
NAME                                   READY   STATUS    RESTARTS   AGE
opencti-0                              2/2     Running   4 (11m ago)   3h20m
```

And if you run `kubectl describe pod opencti-0`, all the containers will have 
as Command ```/charm/bin/pebble```. That's because Pebble is responsible for
the processes startup as explained above.

## High-level overview of a OpenCTI deployment

> Below is a diagram of a basic OpenCTI deployment. It consists of two Juju models
> one VM model for deploying OpenSearch and RabbitMQ, and a k8s model for 
> deploying OpenCTI along with the rest of the dependencies. With the base OpenCTI

Reviewer: ```suggestion
Below is a diagram of a basic OpenCTI deployment. It consists of two Juju models:
one VM model for deploying OpenSearch and RabbitMQ, and a K8s model for 
deploying OpenCTI along with the rest of its dependencies. With the base OpenCTI
```

charm deployed, users can choose from several OpenCTI connector charms to 
integrate with the OpenCTI charm.

```mermaid
C4Container
    System_Ext(k8s-ingress, "Kubernetes Ingress")
    System_Ext(s3, "S3 Compatible Storage")
    Container_Boundary(k8s_model, "K8s Model") {
        Container(ingress, "Nginx Ingress Integrator")
        Container(s3-integrator, "S3 Integrator")
        Container(redis, "Redis")
        Container(opencti, "OpenCTI")
        Container(opencti-connector, "OpenCTI Connector(s)")
        Rel(redis, opencti, "redis")
        Rel(ingress, opencti, "ingress")
        Rel(s3-integrator, opencti , "s3")
        BiRel(opencti, opencti-connector, "opencti-connector")
    }
    Container_Boundary(vm_model, "VM Model") {
        Container(rabbitmq, "RabbitMQ")
        Container(opensearch, "OpenSearch")
    }
    Rel(opensearch, opencti, "opensearch")
    Rel(rabbitmq, opencti, "amqp")
    Rel(ingress, k8s-ingress, "control")
    Rel(s3, s3-integrator, "obtain information")
```

## Charm architecture diagram

> 

Author: Removed the integrations from the system boundary and added a description for the diagram. I also made some minor aesthetic adjustments. Thanks!
